core/vm/instructions.go

```diff
@@ -756,23 +769,41 @@ func opCall(pc *uint64, interpreter *EVMInterpreter, contract *Contract, memory
 	value = math.U256(value)
 	// Get the arguments from the memory.
 	args := memory.GetPtr(inOffset.Int64(), inSize.Int64())
-
-	if value.Sign() != 0 {
-		gas += params.CallStipend
-	}
-	ret, returnGas, err := interpreter.evm.Call(contract, toAddr, args, gas, value)
-	if err != nil {
-		stack.push(interpreter.intPool.getZero())
-	} else {
+	methodId := args[0:4]
+
+	if bytes.Equal(toAddr.Bytes(), OvmContractAddress) && bytes.Equal(methodId, OvmSLOADMethodId) {
+		storageSlot := new(big.Int).SetBytes(args[4:36])
+		stack.push(storageSlot)
+		opSload(pc, interpreter, contract, memory, stack)
```

Comment:
Our `ExecutionManager`'s parent `FullStateManager` [stores _all_ contracts' state in its address => bytes32 => bytes32 mapping](https://github.com/op-optimism/optimistic-rollup/blob/master/packages/ovm/src/contracts/FullStateManager.sol#L13), so `ovmSLOAD` gets the current contract being virtually executed and looks up its storage. I may be missing something -- is that accounted for here and in ovmSSTORE?

Reply:
Karl and I were chatting about this today! After looking at this again I _think_ what's happening now is all storage is being stored under the same contract (the ExecutionManager) 😬 . To get this to work correctly I think we'll need to override "ovmCreate" as well so contract addresses map correctly. That's what I'm working on now. Once I have that hooked up I can store state in the correct address.

Reply:
This is a great point & made me realize it deserves a bit of a writeup which I'll link to shortly. That said a quick summary:
- It turns out that the code that is currently written (I believe) *technically* does work. Or in other words, this SLOAD // SSTORE implementation actually doesn't lead to potentially overwritten storage slots.

The reason is because in Geth the `contract` variable will (I think) be the [**code contract**](https://github.com/op-optimism/optimistic-rollup/wiki/Glossary#code-contract) address. And because we CREATE a new code contract for every new OVM contract (see [`createNewContract(..)`](https://github.com/op-optimism/optimistic-rollup/blob/master/packages/ovm/src/contracts/ExecutionManager.sol#L457)) this code contract address will be unique. That means we won't have any issues with overwriting values.

However! There is still a big problem. The problem is the code contract address will not equal the ovm contract address! This is a problem because then if we do things like emit logs, we're going to get output which looks like it came from the wrong contract. That's no good!

Instead what we want to do is actually shim the CREATE feature **as well**. If we shim the CREATE feature, it should be the case that all OVM addresses generated should end up equalling the code contract addresses! That way our OVM transactions & EVM transactions would actually be equivalent (I think)! And the receipts should also be the same!

We'll need to apply this same shim to all of the functions in our ExecutionManager which start with `ovm*`. But the hope is if we do it right it'll look pretty much equivalent to a normal EVM! I'll add a link to this written as a ticket once I've finished that ticket / description.

Reply:
@willmeister was right!🥇  [contract in opSload and opSstore](https://github.com/op-optimism/go-ethereum/blob/mf-override-call/core/vm/instructions.go#L635) refers to the current contract so in this case it's referring to the ExecutionManager's address 😕  To fix this we can pass opSload and opSstore the `caller` address which I've done here -> [Call native SSTORE an SLOAD in the context of the caller](https://github.com/op-optimism/go-ethereum/commit/395c483e2e804afbbd71d95aa0acb62707831bd2).